ir/value.cpp

```diff
@@ -110,4 +111,37 @@ expr Input::getTyVar() const {
   return expr::mkVar(tyname.c_str(), 2);
 }
 
+
+static string genVectorName(const std::vector<Value *> &vals) {
+  ostringstream ss;
+  ss << "<";
+  for (unsigned idx = 0; idx < vals.size(); ++idx) {
+    ss << *vals[idx];
+    if (idx != vals.size() -1)
+      ss << ", ";
+  }
+  ss << ">";
+  return ss.str();
+}
+
+VectorValue::VectorValue(Type &type, std::vector<Value *> &&vals)
+  : Value(type, genVectorName(vals)), vals(std::move(vals)) {}
+
+void VectorValue::print(ostream &os) const {
+  UNREACHABLE();
+}
+
+StateValue VectorValue::toSMT(State &s) const {
+  expr val;
+  expr not_poison = true;
+
+  for (unsigned idx = 0; idx < vals.size(); ++idx) {
+    auto [v, vs] = vals[idx]->toSMT(s);
+    val = idx == 0 ? v : val.concat(v);
+    not_poison &= vs;
```

Comment:
poison is per element, so we need concat here as well. I think the function just mention can take a vector of StateValues and return a StateValue

Reply:
Hi @nunoplopes , I've fixed this and If you're okay with the code I'll send more code for review (parser and vector operations).